File: packages/totem-sdk/packages/provider-bond/src/policy.ts
import type {
  ProviderBondRegistryState,
  ProviderPolicy,
  PolicyMatch,
  ProviderBondManifest,
  ProviderScore,
} from './types.js';
import { MAX_HEARTBEAT_AGE_MS_DEFAULT } from './constants.js';

export function filterProvidersByPolicy(state: ProviderBondRegistryState, policy: ProviderPolicy): PolicyMatch[] {
  const providers = Object.values(state.providers);
  const now = policy.now ?? Date.now();

  return providers.map((provider) => {
    const reasons: string[] = [];
    const failures: string[] = [];
    const score = state.scores[provider.providerBond.providerId];

    if (policy.serviceType && provider.edgeService.serviceType !== policy.serviceType) {
      failures.push(`Service type ${provider.edgeService.serviceType} does not match required ${policy.serviceType}`);
    }

    if (policy.minScore !== undefined && score) {
      if (score.score < policy.minScore) {
        failures.push(`Score ${score.score} is below minimum ${policy.minScore}`);
      } else {
        reasons.push(`Score ${score.score} meets minimum ${policy.minScore}`);
      }
    }

    if (policy.requireIdentity) {
      if (!provider.providerBond.bondOwnerAddress) {
        failures.push('Identity not declared');
      } else {
        reasons.push('Identity declared');
      }
    }

    if (policy.requireActiveBond) {
      const hasActiveBond = provider.providerBond.bondStack?.some(
        (d) => d.status === 'active'
      );
      if (!hasActiveBond) {
        failures.push('No active bond');
      } else {
        reasons.push('Active bond present');
      }
    }

    if (policy.requireMinimaHardCollateral) {
      const hasMinimaHardCollateral = provider.providerBond.bondStack?.some(
        (d) => d.asset === 'MINIMA' && d.purpose === 'hard-collateral' && d.status === 'active'
      );
      if (!hasMinimaHardCollateral) {
        failures.push('No active MINIMA hard-collateral bond');
      } else {
        reasons.push('MINIMA hard-collateral bond present');
      }
    }

    if (policy.minBondAmount !== undefined) {
      const totalBond = provider.providerBond.bondStack?.reduce(
        (sum, d) => sum + d.amount, 0n
      ) ?? 0n;
      if (totalBond < policy.minBondAmount) {
        failures.push(`Total bond ${totalBond.toString()} is below minimum ${policy.minBondAmount.toString()}`);
      } else {
        reasons.push(`Total bond ${totalBond.toString()} meets minimum ${policy.minBondAmount.toString()}`);
      }
    }

    if (policy.acceptedAssets && policy.acceptedAssets.length > 0) {
      const bondAssets = new Set(provider.providerBond.bondStack?.map((d) => d.asset) ?? []);
      const hasAccepted = policy.acceptedAssets.some((a) => bondAssets.has(a));
      if (!hasAccepted) {
        failures.push('No bond in accepted assets');
      } else {
        reasons.push('Bond asset accepted');
      }
    }

    if (policy.acceptedPurposes && policy.acceptedPurposes.length > 0) {
      const bondPurposes = new Set(provider.providerBond.bondStack?.map((d) => d.purpose) ?? []);
      const hasAccepted = policy.acceptedPurposes.some((p) => bondPurposes.has(p));
      if (!hasAccepted) {
        failures.push('No bond with accepted purpose');
      } else {
        reasons.push('Bond purpose accepted');
      }
    }

    if (policy.maxIncidentSeverity) {
      const providerIncidents = state.incidents[provider.providerBond.providerId] ?? [];
      const severityOrder: Record<string, number> = { low: 0, medium: 1, high: 2, critical: 3 };
      const maxSeen = providerIncidents.reduce((max, i) => {
        const s = severityOrder[i.severity] ?? 0;
        return s > max ? s : max;
      }, 0);
      const policyMax = severityOrder[policy.maxIncidentSeverity] ?? 0;
      if (maxSeen > policyMax) {
        failures.push(`Incident severity exceeds policy maximum`);
      } else {
        reasons.push('Incident severity within policy limits');
      }
    }

    if (policy.maxHeartbeatAgeMs !== undefined) {
      const providerProbes = state.probes[provider.providerBond.providerId] ?? [];
      const heartbeats = providerProbes.filter((p) => p.type === 'heartbeat');
      const lastHeartbeat = heartbeats.length > 0
        ? heartbeats.reduce((max, h) => h.observedAt > max ? h.observedAt : max, 0)
        : 0;
      if (now - lastHeartbeat > policy.maxHeartbeatAgeMs) {
        failures.push(`Last heartbeat too old (${now - lastHeartbeat}ms > ${policy.maxHeartbeatAgeMs}ms)`);
      } else if (lastHeartbeat > 0) {
        reasons.push('Heartbeat is recent');
      }
    }

    return {
      providerId: provider.providerBond.providerId,
      provider,
      matched: failures.length === 0,
      score,
      reasons,
      failures,
    };
  });
}

export function rankProvidersByPolicy(matches: PolicyMatch[]): PolicyMatch[] {
  return [...matches].sort((a, b) => {
    if (a.matched && !b.matched) return -1;
    if (!a.matched && b.matched) return 1;
    const scoreA = a.score?.score ?? 0;
    const scoreB = b.score?.score ?? 0;
    return scoreB - scoreA;
  });
}

export function explainProviderPolicyMatch(match: PolicyMatch): string[] {
  const lines: string[] = [];
  lines.push(`Provider: ${match.providerId}`);
  lines.push(`Matched: ${match.matched}`);
  if (match.score) {
    lines.push(`Score: ${match.score.score} (${match.score.recommendation})`);
  }
  if (match.reasons.length > 0) {
    lines.push('Passed checks:');
    for (const r of match.reasons) lines.push(`  + ${r}`);
  }
  if (match.failures.length > 0) {
    lines.push('Failed checks:');
    for (const f of match.failures) lines.push(`  - ${f}`);
  }
  return lines;
}

File: packages/totem-sdk/packages/provider-bond/rust/src/policy.rs
use crate::scoring::compute_recommendation;
use crate::types::*;

pub fn filter_providers_by_policy(
    provider: &ProviderBondManifest,
    policy: &ProviderPolicy,
    score: Option<&ProviderScore>,
    probes: &[ProbeResult],
    incidents: &[IncidentRecord],
) -> PolicyMatch {
    let now = policy.now.unwrap_or(0);
    let mut reasons: Vec<String> = Vec::new();
    let mut failures: Vec<String> = Vec::new();

    if let Some(ref min_score) = policy.min_score {
        if let Some(s) = score {
            if s.score < *min_score {
                failures.push(format!("Score {} is below minimum {}", s.score, min_score));
            } else {
                reasons.push(format!("Score {} meets minimum {}", s.score, min_score));
            }
        }
    }

    if policy.require_identity.unwrap_or(false) {
        if provider.provider_bond.bond_owner_address.is_none() {
            failures.push("Identity not declared".to_string());
        } else {
            reasons.push("Identity declared".to_string());
        }
    }

    if policy.require_active_bond.unwrap_or(false) {
        let has_active = provider.provider_bond.bond_stack.as_ref().map_or(false, |s| {
            s.iter().any(|d| d.status == "active")
        });
        if !has_active {
            failures.push("No active bond".to_string());
        } else {
            reasons.push("Active bond present".to_string());
        }
    }

    if policy.require_minima_hard_collateral.unwrap_or(false) {
        let has_minima = provider.provider_bond.bond_stack.as_ref().map_or(false, |s| {
            s.iter().any(|d| d.asset == "MINIMA" && d.purpose == "hard-collateral" && d.status == "active")
        });
        if !has_minima {
            failures.push("No active MINIMA hard-collateral bond".to_string());
        } else {
            reasons.push("MINIMA hard-collateral bond present".to_string());
        }
    }

    if let Some(ref min_amount) = policy.min_bond_amount {
        let total: u128 = provider.provider_bond.bond_stack.as_ref().map_or(0, |s| {
            s.iter().filter_map(|d| d.amount.parse::<u128>().ok()).sum()
        });
        let min: u128 = min_amount.parse().unwrap_or(0);
        if total < min {
            failures.push(format!("Total bond {} is below minimum {}", total, min));
        } else {
            reasons.push(format!("Total bond {} meets minimum {}", total, min));
        }
    }

    if let Some(ref accepted_assets) = policy.accepted_assets {
        if !accepted_assets.is_empty() {
            let bond_assets: std::collections::HashSet<&str> = provider.provider_bond.bond_stack.as_ref().map_or(
                std::collections::HashSet::new(),
                |s| s.iter().map(|d| d.asset.as_str()).collect(),
            );
            let has_accepted = accepted_assets.iter().any(|a| bond_assets.contains(a.as_str()));
            if !has_accepted {
                failures.push("No bond in accepted assets".to_string());
            } else {
                reasons.push("Bond asset accepted".to_string());
            }
        }
    }

    if let Some(ref accepted_purposes) = policy.accepted_purposes {
        if !accepted_purposes.is_empty() {
            let bond_purposes: std::collections::HashSet<&str> = provider.provider_bond.bond_stack.as_ref().map_or(
                std::collections::HashSet::new(),
                |s| s.iter().map(|d| d.purpose.as_str()).collect(),
            );
            let has_accepted = accepted_purposes.iter().any(|p| bond_purposes.contains(p.as_str()));
            if !has_accepted {
                failures.push("No bond with accepted purpose".to_string());
            } else {
                reasons.push("Bond purpose accepted".to_string());
            }
        }
    }

    if let Some(ref max_severity) = policy.max_incident_severity {
        let max_seen = incidents.iter().map(|i| severity_value(&i.severity)).max().unwrap_or(0);
        let policy_max = severity_value(max_severity);
        if max_seen > policy_max {
            failures.push("Incident severity exceeds policy maximum".to_string());
        } else {
            reasons.push("Incident severity within policy limits".to_string());
        }
    }

    if let Some(max_age) = policy.max_heartbeat_age_ms {
        let heartbeats: Vec<&ProbeResult> = probes.iter().filter(|p| p.probe_type == "heartbeat").collect();
        let last_heartbeat = heartbeats.iter().map(|h| h.observed_at).max().unwrap_or(0);
        if now > 0 && last_heartbeat > 0 && now - last_heartbeat > max_age {
            failures.push(format!("Last heartbeat too old ({}ms > {}ms)", now - last_heartbeat, max_age));
        } else if last_heartbeat > 0 {
            reasons.push("Heartbeat is recent".to_string());
        }
    }

    PolicyMatch {
        provider_id: provider.provider_bond.provider_id.clone(),
        matched: failures.is_empty(),
        reasons,
        failures,
    }
}

fn severity_value(severity: &str) -> u32 {
    match severity {
        "low" => 0,
        "medium" => 1,
        "high" => 2,
        "critical" => 3,
        _ => 0,
    }
}
